packages/berry-pnpify/sources/NodeModulesFS.ts

```diff
@@ -0,0 +1,253 @@
+  }
+
+  getBaseFs() {
+    return this.baseFs;
+  }
+
+  private resolveFilePath(p: string): string {
+    const pnpPath = this.pathResolver.resolvePath(p);
+    if (!pnpPath.resolvedPath) {
+      throw new Error(`ENOENT: no such file or directory, stat '${p}'`)
+    } else {
+      return pnpPath.resolvedPath;
+    }
+  }
+
+  private resolveStatPath(p: string): string {
+    const pnpPath = this.pathResolver.resolvePath(p);
+    if (!pnpPath.resolvedPath) {
+      throw new Error(`ENOENT: no such file or directory, stat '${p}'`)
+    } else {
+      return pnpPath.statPath || pnpPath.resolvedPath;
+    }
+  }
+
+  async openPromise(p: string, flags: string, mode?: number) {
+    return await this.baseFs.openPromise(this.resolveFilePath(p), flags, mode);
+  }
+
+  openSync(p: string, flags: string, mode?: number) {
+    return this.baseFs.openSync(this.resolveFilePath(p), flags, mode);
+  }
+
+  async closePromise(fd: number) {
+    await this.baseFs.closePromise(fd);
+  }
+
+  closeSync(fd: number) {
+    this.baseFs.closeSync(fd);
+  }
+
+  createReadStream(p: string, opts?: CreateReadStreamOptions) {
+    return this.baseFs.createReadStream(this.resolveFilePath(p), opts);
+  }
+
+  createWriteStream(p: string, opts?: CreateWriteStreamOptions) {
+    return this.baseFs.createWriteStream(this.resolveFilePath(p), opts);
+  }
+
+  async realpathPromise(p: string) {
+    return await this.baseFs.realpathPromise(this.resolveFilePath(p));
+  }
+
+  realpathSync(p: string) {
+    return this.baseFs.realpathSync(this.resolveFilePath(p));
+  }
+
+  async existsPromise(p: string) {
+    const pnpPath = this.pathResolver.resolvePath(p);
+    if (!pnpPath.resolvedPath) {
+      return false;
+    } else {
+      return await this.baseFs.existsPromise(pnpPath.resolvedPath);
+    }
+  }
+
+  existsSync(p: string) {
+    const pnpPath = this.pathResolver.resolvePath(p);
+    if (!pnpPath.resolvedPath) {
+      return false;
+    } else {
+      return this.baseFs.existsSync(pnpPath.resolvedPath);
+    }
+  }
+
+  async accessPromise(p: string, mode?: number) {
+    return await this.baseFs.accessPromise(this.resolveStatPath(p), mode);
+  }
+
+  accessSync(p: string, mode?: number) {
+    return this.baseFs.accessSync(this.resolveStatPath(p), mode);
+  }
+
+  async statPromise(p: string) {
+    return await this.baseFs.statPromise(this.resolveStatPath(p));
+  }
+
+  statSync(p: string) {
+    return this.baseFs.statSync(this.resolveStatPath(p));
+  }
+
+  async lstatPromise(p: string) {
+    return await this.baseFs.lstatPromise(this.resolveStatPath(p));
+  }
+
+  lstatSync(p: string) {
+    return this.baseFs.lstatSync(this.resolveStatPath(p));
+  }
+
+  async chmodPromise(p: string, mask: number) {
+    return await this.baseFs.chmodPromise(p, mask);
+  }
+
+  chmodSync(p: string, mask: number) {
+    return this.baseFs.chmodSync(p, mask);
+  }
+
+  async renamePromise(oldP: string, newP: string) {
+    return await this.baseFs.renamePromise(oldP, newP);
```

Comment:
Thinking about it, maybe we should make this filesystem read-only and throw exceptions and all write accesses within the `node_modules`, what do you think?

Reply:
@arcanis Yes, sounds good to me